sky/provision/vast/utils.py

```diff
@@ -0,0 +1,103 @@
+# pylint: disable=assignment-from-no-return
+"""Vast library wrapper for SkyPilot."""
+from typing import Any, Dict, List, Optional
+
+from sky import sky_logging
+from sky.adaptors import vast
+
+logger = sky_logging.init_logger(__name__)
+
+
+def list_instances() -> Dict[str, Dict[str, Any]]:
+    """Lists instances associated with API key."""
+    instances = vast.vast().show_instances()
+
+    instance_dict: Dict[str, Dict[str, Any]] = {}
+    for instance in instances:
+        instance['id'] = str(instance['id'])
+        info = instance.copy()
+
+        if isinstance(instance['actual_status'], str):
+            info['status'] = instance['actual_status'].upper()
+        else:
+            info['status'] = 'UNKNOWN'
+        info['name'] = instance['label']
+
+        instance_dict[instance['id']] = info
+
+    return instance_dict
+
+
+def launch(name: str, instance_type: str, region: str, disk_size: int,
+           image_name: str, ports: Optional[List[int]], public_key: str) -> str:
+    """Launches an instance with the given parameters.
+
+    Converts the instance_type to the Vast GPU name, finds the specs for the
+    GPU, and launches the instance.
+    """
+    del ports
+    del public_key
+
+    gpu_name = instance_type.split('-')[1].replace('_', ' ')
+    num_gpus = int(instance_type.split('-')[0].replace('x', ''))
+
+    query = ' '.join([
+        f'geolocation="{region[-2:]}"',
+        f'disk_space>={disk_size}',
+        f'num_gpus={num_gpus}',
+        f'gpu_name="{gpu_name}"',
+    ])
+
+    instance_list = vast.vast().search_offers(query=query)
```

Comment:
Is this instance or offers (e.g. instance types)?

Reply:
we're somewhat different. Most providers have a set of instance types but we technically offer snowflakes and entrust our users to abstract away to their needs. I was just talking about how this imposes extra work on our customers yesterday so we may be introducing categories but as of now, all classification is client-side through an expressive querying syntax.